### PR TITLE
[LoRA] Add support for pinning lora adapters in the LRU cache

### PR DESCRIPTION
**Pull request by rohithkrn**
Adds support to pin certain lora adapters in the LRUCache. This helps certain latency sensitive adapters to be present in GPU memory without being evicted.

### CODE REVIEW

File: vllm/lora/models.py [hunk 1]
```diff
@@ -525,6 +525,26 @@ def remove_lora(self, lora_id: int) -> bool:
...
+        """Pin a LoRAModel in the manager cache."""
+        self._pin_lora_in_cpu_cache(lora_id)
+        self._pin_lora_in_gpu_cache(lora_id)
+        return True
+
+    def _pin_lora_in_cpu_cache(self, lora_id: int):
+        try:
+            self._registered_loras.pin(lora_id)
+        except ValueError as err:
+            raise ValueError(f"Pinning failed. \
```
Comment (Yard1) [inline vllm/lora/models.py:538]: avoid \ formatting

File: vllm/utils.py [hunk 2]
```diff
@@ -101,14 +102,31 @@ def put(self, key: Hashable, value: T) -> None:
         self.cache.move_to_end(key)
         self._remove_old_if_needed()
 
+    def pin(self, key: Hashable) -> None:
```
Comment (Yard1) [inline vllm/utils.py:105]: can you add a docstring explaining what "pin" means, to make it easier for future contributors?

File: vllm/utils.py [hunk 3]
```diff
@@ -101,14 +102,31 @@ def put(self, key: Hashable, value: T) -> None:
...
-        key, value = self.cache.popitem(last=False)
-        self._on_remove(key, value)
+
+        if not remove_pinned:
+            if all(key in self.pinned_items for key in self.cache):
+                raise RuntimeError("All items are pinned, \
+                                   cannot remove oldest from the cache.")
+            # pop the oldest item in the cache that is not pinned
+            lru_key = next(key for key in self.cache
+                           if key not in self.pinned_items)
```
Comment (Yard1) [inline vllm/utils.py:126]:
```suggestion
            # pop the oldest item in the cache that is not pinned
            lru_key = next((key for key in self.cache
                           if key not in self.pinned_items), ALL_PINNED_SENTINEL)
            if lru_key is ALL_PINNED_SENTINEL:
                raise RuntimeError("All items are pinned, "
                                   "cannot remove oldest from the cache.")
```

and somewhere in the file
```
class _Sentinel:
    ...

ALL_PINNED_SENTINEL = _Sentinel()
```

**Review by Yard1 — COMMENTED** [review]
overall looks good to me

File: vllm/lora/models.py (hunk 1 above)
Comment (rohithkrn) [reply vllm/lora/models.py]: Updated.

File: vllm/utils.py (hunk 2 above)
Comment (rohithkrn) [reply vllm/utils.py:112]: Added.

File: vllm/utils.py (hunk 3 above)
Comment (rohithkrn) [reply vllm/utils.py]: thanks for the suggestion, updated.

**Comment by rohithkrn** [comment]
@Yard1 ready for review. thanks

**Comment by rohithkrn** [comment]
as for the failing CI jobs, they are related to lint error about function return type. Other similar functions have the same signature, hence did not change.

**Comment by rohithkrn** [comment]
@Yard1 gentle reminder for review.

**Comment by Yard1** [comment]
@rohithkrn we need to get the lint to pass. If there's no way to actually fix the lint issues, we should use `noqa` comments to make the linter ignore specific lines. Otherwise looks good to me.

**Comment by rohithkrn** [comment]
@Yard1 Passed CI

**Review by Yard1 — APPROVED** [review]
Thanks!

**Pull request merged**